plugins/Unity/Launcher/xdgwatcher.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2019 UBports project.
+ * Author(s): Marius Gripsgard <marius@ubports.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdgwatcher.h"
+
+#include <QDebug>
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QStandardPaths>
+#include <QTextStream>
+
+XdgWatcher::XdgWatcher(QObject* parent)
+    : QObject(parent),
+      m_watcher(new QFileSystemWatcher(this))
+{
+    connect(m_watcher, &QFileSystemWatcher::directoryChanged, this, &XdgWatcher::onDirectoryChanged);
+    connect(m_watcher, &QFileSystemWatcher::fileChanged, this, &XdgWatcher::onFileChanged);
+
+    const auto paths = QStandardPaths::standardLocations(QStandardPaths::ApplicationsLocation);
+    for (const auto &path: paths) {
+        const auto qdir = QDir(path);
+        if (!qdir.exists()) {
+            continue;
+        }
+
+        // Add the path itself to watch for newly added apps
+        m_watcher->addPath(path);
+
+        // Add watcher for eatch app to watch for changes
+        const auto files = qdir.entryInfoList(QDir::Files);
+        for (const auto &file: files) {
+            if (file.suffix() == "desktop") {
+                m_watcher->addPath(file.absoluteFilePath());
+            }
+        }
+    }
+}
+
+// "Ubuntu style" appID is filename without versionNumber after last "_"
+const QString XdgWatcher::stripAppIdVersion(const QString rawAppID) {
+    auto appIdComponents = rawAppID.split("_");
+    appIdComponents.removeLast();
+    return appIdComponents.join("_");
+}
+
+// Standard appID see:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+const QString XdgWatcher::toStandardAppId(const QFileInfo fileInfo) const {
+    const auto paths = QStandardPaths::standardLocations(QStandardPaths::ApplicationsLocation);
+    for (const auto &path: paths) {
+        if (fileInfo.absolutePath() == path) {
+            break;
+        }
+        if (fileInfo.absolutePath().contains(path)) {
+            auto fileStr = fileInfo.absoluteFilePath();
+            fileStr.replace(path, "");
+            fileStr.replace("/", "-");
+            fileStr.replace(".desktop", "");
+            return fileStr;
+        }
+    }
+    return fileInfo.completeBaseName();
+}
+
+const QString XdgWatcher::toAppId(const QFileInfo fileInfo) const {
+    QFile qFile(fileInfo.absoluteFilePath());
+    qFile.open(QIODevice::ReadOnly);
```

Comment:
Actually, I think we can just get rid of the whole need to open the file and check for `X-Ubuntu-Application-ID` entirely, and only use `toStandardAppId()` to get the app ID, since click hooks place the files as the correct `${APP_ID}.desktop` filename. Then we can just always strip the version.

Reply:
Yeah i guess that should also work just fine, since we use to remove the file anyway.

Reply:
This has still not been updated?

Reply:
Thinking about this, we may still improperly strip for example org_kde_kate, so we would still need to open the file to check

Reply:
I don't think this is a real concern. KDE does not install applications in this manner.

Reply:
kde was just an example, but one app that does do this on my local system is `krita_jpg_jpeg.desktop`